### ISSUE TITLE
Cannot send packet inside notify* functions

### ISSUE THREAD

**Issue by positron96** [mrrwa/LocoNet2#4]
Hello! 
I'm trying to use and modify this library for my purposes and noticed that I cannot send packets from inside `notify...` methods, this makes responding to requests problematic. The issue is due to actual sending bits and receiving bits happens in the same freertos task and  `while(lnState == LN_ST_TX)` loop inside `sendLocoNetPacketTry`  basically blocks the task it's running on. When one calls `send` method from `notify...` function, that loops basically deadlocks the sending task.

I think it would be better to move actual bit sending (and receiving) to timer functions and only handle receiving of bytes in the task. This way, waiting for send to complete won't block sending which is done in timers.

Also, what's the development status of this library? I've noticed that it hasn't been updated for a while.

**Comment by kiwi64ajs** [comment]
Hi Pavel,

> On 23/05/2020, at 9:51 PM, Pavel Melnikov <notifications@github.com> wrote:
> Hello!
> I'm trying to use and modify this library for my purposes and noticed that I cannot send packets from inside notify... methods, this makes responding to requests problematic. The issue is due to actual sending bits and receiving bits happens in the same freertos task and while(lnState == LN_ST_TX) loop inside sendLocoNetPacketTry basically blocks the task it's running on. When one calls send method from notify... function, that loops basically deadlocks the sending task.
> 
> I think it would be better to move actual bit sending (and receiving) to timer functions and only handle receiving of bytes in the task. This way, waiting for send to complete won't block sending which is done in timers.
> 
Yeah, a fundamental issue with LocoNet is that you can’t imediately send a response because each packet must have a CD-Backoff delay and even then you can’t be sure that you’re not blocked by another node sending.

You need to have some sort of deferred processing required flag or queue etc to handle this sort of thing.
> Also, what's the development status of this library? I've noticed that it hasn't been updated for a while.
> 
I forked it from the AVR based one as I could see I’d painted myself into a corner with that one, but didn’t complete the work. I’ve continued to do bug fixes to the original but not pasted all of them into the LocoNet2 library. A few others have done work on it - check the forks in GitHub.

HTH

Alex Shepherd

**Comment by positron96** [comment]
Hi Alex. Thanks for the response. 

Actually what I was pointing out is not the inherent issue with LocoNet protocol but the way it is implemented. The problem lies in that the task that sends bits to network also does receiving and processing bits from network. So when a packet arrives, its notify function is called from that task. If there is a `send` inside notify function, the task comes to a busy loop that waits for data to be sent. But that is never sent because sending is done by that same task. If, for example, sending was implemented in another task or by timer interrupt (as it is done in AVR version), there would be no issue even considering CD backoffs and stuff - sending would eventually succeed and the task would continue.

I found only one fork that had some commits. The timer+bitbanging code does not seem to work (or I've messed with its timings somehow), but there is an implementation via UART driver that kinda works but does not allow to invert tx pin while not inverting rx pin - which most DIY LocoNet schematics require. On the good side, it does not seem to be deadlocking in the described case, though I have not tested it yet.

Anyway, thanks for your good work with the libraries. I'll keep poking around in the library and its fork, trying to fit it to my hardware and software requirements, probably making some changes in the process. Do you accept pull requests though?
